### PR TITLE
[DPP-589] Add CLI flag to select minimum enabled TLS version

### PR DESCRIPTION
**Pull request by pbatko-da**
- Add support for specifying either 1.2 or 1.3 as minimum TLS versions for ledger api server.
- Log enabled protocols (~TLS versions) and cipher suites at server and client startup.
- Add integration tests against Sandbox-classic and Sandbox

CHANGELOG_BEGIN
Sandbox: Add CLI flag to select minimum enabled TLS version for participant server.
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala [hunk 1]
```diff
@@ -195,6 +195,15 @@ class CommonCliBase(name: LedgerName) {
           )
         )
 
+      opt[TlsVersion]("tls-version")
```
Comment (mziolekda) [inline ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:198]: I suggest calling that flag "min-tls-version" to drive the point home that it is not just a single version that is supported.

File: ledger/daml-on-sql/README.rst [hunk 2]
```diff
@@ -226,6 +226,9 @@ which should serve decryption details as a JSON document like so:
 
 Sample command to start a server with private key encrypted: ``java -jar daml-on-sql-<version>.jar --pem server.pem.enc --crt server.crt --tls-secrets-url http://localhost:8080``.
 
+To specify minimum TLS version for the server, add either ``--tls-version 1.2`` or ``--tls-version 1.3``
```
Comment (mziolekda) [inline ledger/daml-on-sql/README.rst:229]:
```suggestion
To specify minimum TLS version for the server, add ``--tls-version <version>`` to the server invocation
```

Comment (pbatko-da) [reply ledger/daml-on-sql/README.rst:229]: Ok

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala (hunk 1 above)
Comment (pbatko-da) [reply ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:198]: Ok

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala [hunk 3]
```diff
@@ -3,70 +3,166 @@
...
+      s"TLS $who: Supported cipher suites: ${tlsInfo.supportedCipherSuites.mkString(", ")}."
+    )
+    logger.info(s"TLS $who: Enabled cipher suites: ${tlsInfo.enabledCipherSuites.mkString(", ")}.")
+  }
 
   /** This is a side-effecting method. It modifies JVM TLS properties according to the TLS configuration. */
   def setJvmTlsProperties(): Unit =
     if (enabled && enableCertRevocationChecking) OcspProperties.enableOcsp()
 
+  private[tls] def protocolsNames(tlsInfo: TlsInfo): java.lang.Iterable[String] = {
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala:147]:
```suggestion
  private[tls] def filterSupportedProtocols(tlsInfo: TlsInfo): java.lang.Iterable[String] = {
```
Use a verb in the function name that describes what is being done to the list of the possible protocols

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala [hunk 4]
```diff
@@ -3,70 +3,166 @@
...
+      .build()
+  }
+
+  private[tls] def logTlsProtocolsAndCipherSuites(
+      sslContext: SslContext,
+      isServer: Boolean,
+  ): Unit = {
+    val who = if (isServer) "server" else "client"
+    val tlsInfo = TlsInfo.fromSslContext(sslContext)
+    logger.info(s"TLS $who: Enabled.")
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala:134]: I would rephrase this so that we say things such as:
Server TLS - enabled
Server TLS - supported protocols: <prot>, <prot>, <prot>

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala [hunk 5]
```diff
@@ -46,6 +64,43 @@ class TlsConfigurationTest extends AnyWordSpec with Matchers with BeforeAndAfter
   }
 
   "TlsConfiguration" should {
+
+    "configure server with TLS protocol versions" which {
+      "is TLS 1.3" in {
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala:69]:
```suggestion
      "is 1.3" in {
```
We already use the TLS in this sentence once

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala [hunk 6]
```diff
@@ -46,6 +64,43 @@ class TlsConfigurationTest extends AnyWordSpec with Matchers with BeforeAndAfter
...
+        getServerEnabledProtocols(Some(TlsVersion.V1_3)) shouldBe Seq("SSLv2Hello", "TLSv1.3")
+      }
+      "is TLS 1.2" in {
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala:72]:
```suggestion
      "is 1.2" in {
```

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala [hunk 7]
```diff
@@ -46,6 +64,43 @@ class TlsConfigurationTest extends AnyWordSpec with Matchers with BeforeAndAfter
...
+          "TLSv1",
+          "TLSv1.1",
+          "TLSv1.2",
+          "TLSv1.3",
+        )
+      }
+    }
+
+    "configure client with TLS protocol versions" which {
+      "is TLS 1.3" in {
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala]:
```suggestion
      "is 1.3" in {
```

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala [hunk 8]
```diff
@@ -46,6 +64,43 @@ class TlsConfigurationTest extends AnyWordSpec with Matchers with BeforeAndAfter
...
+        getClientEnabledProtocols(Some(TlsVersion.V1_3)) shouldBe Seq("SSLv2Hello", "TLSv1.3")
+      }
+      "is TLS 1.2" in {
```
Comment (mziolekda) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala]:
```suggestion
      "is 1.2" in {
```

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/BaseTlsServerIT.scala [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+import java.io.File
+import scala.concurrent.Future
+
+abstract class BaseTlsServerIT(minimumServerProtocolVersion: Option[TlsVersion])
+    extends AsyncWordSpec
+    with AbstractSandboxFixture
+    with SuiteResourceManagementAroundAll {
+
+  minimumServerProtocolVersion match {
+    case Some(TlsVersion.V1_3) =>
```
Comment (mziolekda) [inline ledger/sandbox-common/src/test/lib/scala/platform/sandbox/BaseTlsServerIT.scala:33]: I like those tests

**Review by mziolekda — COMMENTED** [review]
Overall, very solid PR. I have requested a couple more changes.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala [hunk 10]
```diff
@@ -3,70 +3,166 @@
...
+      Some(sslContext)
+    } else None
   }
 
   /** If enabled and all required fields are present, it returns an SslContext suitable for server usage */
   def server: Option[SslContext] =
     if (enabled) {
-      scala.util.Using.resources(
+
+      val tlsInfo = scala.util.Using.resources(
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala:79]: Ok

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala (hunk 3 above)
Comment (pbatko-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala:147]: Ok

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala (hunk 4 above)
Comment (pbatko-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/tls/TlsConfiguration.scala:134]: Ok

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala (hunk 6 above)
Comment (pbatko-da) [reply ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala:72]: Ok

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala (hunk 5 above)
Comment (pbatko-da) [reply ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/tls/TlsConfigurationTest.scala:69]: Ok

**Review by mziolekda — APPROVED** [review]
Thank you, nicely done!

**Review by kamil-da — APPROVED** [review]
